### PR TITLE
Move coreapi tests to the interface

### PR DESCRIPTION
**Pull request by magik6k**
Needs to be done before interface extraction

### CODE REVIEW

**Comment by eingenito** [comment]
Hey @magik6k, I'm taking a look at this and I can follow what you did mechanically, but I don't understand why you did it? I have some questions; I apologize if they don't make sense.

* What's the extraction you're referring to in the initial comment?
* What's the motivation for tracking contexts as you do here? Has this been an issue in CI or something?
* I think this will prevent us from running any of these tests individually? Is that a big deal?

Any other context you could provide quickly would be great. I get that I might just be the wrong reviewer but you mentioned this PR about two weeks ago and I wanted to get you some feedback. Thanks.

File: core/coreapi/interface/tests/api.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+		t.Run("Pin", tp.TestPin)
+		t.Run("PubSub", tp.TestPubSub)
+		t.Run("Unixfs", tp.TestUnixfs)
+
+		apis <- -1
+		t.Run("TestsCancelCtx", func(t *testing.T) {
+			select {
+			case <-zeroRunning:
+			case <-time.After(time.Second):
+				t.Errorf("%d node(s) not closed", running)
```
Comment (eingenito) [inline core/coreapi/interface/tests/api.go]: nit: we're not really tracking nodes here right? More like constructed swarms.

**Comment by magik6k** [comment]
> What's the extraction you're referring to in the initial comment?

https://github.com/ipfs/interface-ipfs-core/issues/66, basically we don't want external implementations of coreapi to depend on go-ipfs repo

> What's the motivation for tracking contexts as you do here? Has this been an issue in CI or something?

[go-ipfs-http-api](https://github.com/ipfs/go-ipfs-http-client/pull/1) uses IPTB to spawn nodes and relies on contexts to kill those nodes.

> I think this will prevent us from running any of these tests individually? Is that a big deal?

Running all coreapi tests takes about 3-4s for the `core/coreapi` implementation on my PC. It will probably be much slower with http-api+IPTB, but filtering with `-run` still works just fine (e.g. `go test github.com/ipfs/go-ipfs/core/coreapi/test -run='TestIface/Object/TestObjectPut' -v`)

File: core/coreapi/interface/tests/block.go [hunk 2]
```diff
@@ -103,16 +116,17 @@ func TestBlockGet(t *testing.T) {
...
+	defer cancel()
+	api, err := tp.makeAPI(ctx)
 	if err != nil {
 		t.Error(err)
 	}
 
 	res, err := api.Block().Put(ctx, strings.NewReader(`Hello`))
 	if err != nil {
-		t.Error(err)
+		t.Fatal(err)
```
Comment (eingenito) [inline core/coreapi/interface/tests/block.go]: What's the pattern for the replacement of Error with Fatal? Are we testing the first use of the CoreAPI generally?

File: core/coreapi/interface/tests/key.go [hunk 3]
```diff
@@ -125,11 +154,12 @@ func TestGenerateSize(t *testing.T) {
 	}
 }
 
-func TestGenerateType(t *testing.T) {
-	ctx := context.Background()
+func (tp *provider) TestGenerateType(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (eingenito) [inline core/coreapi/interface/tests/key.go]: These could go below skip?

File: core/coreapi/interface/tests/block.go (hunk 2 above)
Comment (eingenito) [reply core/coreapi/interface/tests/block.go]: I think I understand - it's an existing inconsistency and you didn't want to change them wholesale.

File: core/coreapi/interface/tests/key.go (hunk 3 above)
Comment (Stebalien) [reply core/coreapi/interface/tests/key.go]: (yes but I didn't see that comment before I merged... oh well, it doesn't really matter much).

**Pull request merged**